File: Python/easy/234. Palindrome Linked List.py
"""
Given a singly linked list, determine if it is a palindrome.
Example 1:
Input: 1->2
Output: false
Example 2:
Input: 1->2->2->1
Output: true
"""
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# easiest solution, O(n) space
class Solution:
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        curNode = head
        valList = []
        while curNode != None:
            valList.append(curNode.val)
            curNode = curNode.next
        for i in range(int(len(valList) / 2)):
            if valList[i] != valList[-i - 1]:
                return False
        return True

# change the linked list, O(1) space
class Solution:
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        fast = slow = head
        while fast != None and fast.next != None:
            fast = fast.next.next
            slow = slow.next
        preVal = None
        while slow != None:
            temp = slow.next
            slow.next = preVal
            preVal = slow
            slow = temp
        curNode = head
        while  preVal != None:
            if preVal.val != curNode.val:
                return False
            curNode = curNode.next
            preVal = preVal.next
        return True

# change the linked list but restore it at the end, O(1) space
class Solution:
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        fast = slow = head
        while fast != None and fast.next != None:
            fast = fast.next.next
            slow = slow.next
        preVal = None
        while slow != None:
            temp = slow.next
            slow.next = preVal
            preVal = slow
            slow = temp
        curNode = head
        nextVal = None
        while preVal != None:
            if preVal.val != curNode.val:
                return False
            curNode = curNode.next
            temp = preVal.next
            preVal.next = nextVal
            nextVal = preVal
            preVal = temp
        return True

File: Java/easy/234. Palindrome Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if (head == null) return true;
        ListNode fastNode = head;
        ListNode slowNode = head;
        while (fastNode.next != null && fastNode.next.next != null) {
            fastNode = fastNode.next.next;
            slowNode = slowNode.next;
        }
        ListNode prevNode = null;
        ListNode curNode = slowNode.next;
        while (curNode != null) {
            ListNode tempNode = curNode.next;
            curNode.next = prevNode;
            prevNode = curNode;
            curNode = tempNode;
        }
        curNode  = head;
        while (prevNode != null) {
            if (prevNode.val != curNode.val) return false;
            curNode = curNode.next;
            prevNode = prevNode.next;
        }
        return true;
    }
}